### PR TITLE
[IMP] *: add props validation

### PR DESCRIPTION
**Pull request by pro-odoo**
With this revision, props of all our components are now validated.

## Description:

description of this task, what is implemented and why it is implemented that way.

Odoo task ID : [TASK_ID](https://www.odoo.com/web#id=TASK_ID&action=333&active_id=2328&model=project.task&view_type=form&cids=1&menu_id=4720)

## review checklist

- [ ] feature is organized in plugin, or UI components
- [ ] support of duplicate sheet (deep copy)
- [ ] in model/core: ranges are Range object, and can be adapted (adaptRanges)
- [ ] in model/UI: ranges are strings (to show the user)
- [ ] undo-able commands (uses this.history.update)
- [ ] multiuser-able commands (has inverse commands and transformations where needed)
- [ ] new/updated/removed commands are documented
- [ ] exportable in excel
- [ ] translations (\_lt("qmsdf %s", abc))
- [ ] unit tested
- [ ] clean commented code
- [ ] track breaking changes
- [ ] doc is rebuild (npm run doc)
- [ ] status is correct in Odoo

### CODE REVIEW

File: src/components/figures/figure_chart/figure_chart.ts [hunk 1]
```diff
@@ -135,3 +135,9 @@ export class ChartFigure extends Component<Props, SpreadsheetChildEnv> {
     return component;
   }
 }
+
+ChartFigure.props = {
+  figure: Object,
+  sidePanelIsOpen: Boolean,
```
Comment (hokolomopo) [inline src/components/figures/figure_chart/figure_chart.ts:141]: Same comment as above

File: src/components/figures/figure_container/figure_container.ts [hunk 2]
```diff
@@ -31,4 +31,9 @@ export class FiguresContainer extends Component<Props, SpreadsheetChildEnv> {
   }
 }
 
+FiguresContainer.props = {
+  sidePanelIsOpen: Boolean,
```
Comment (hokolomopo) [inline src/components/figures/figure_container/figure_container.ts:35]: Same comment as above

File: src/components/filters/filter_menu/filter_menu.ts [hunk 3]
```diff
@@ -310,6 +310,7 @@ export class FilterMenu extends Component<Props, SpreadsheetChildEnv> {
     this.props.onClosed?.();
   }
 }
+FilterMenu.props = { filterPosition: Object, onClosed: { type: Function, optional: true } };
```
Comment (hokolomopo) [inline src/components/filters/filter_menu/filter_menu.ts:313]: Prettier ? Why is this one in one line, where `FilterIconsOverlay.props` which is shorter is on mutliple lines

File: src/components/filters/filter_menu_item/filter_menu_value_item.ts [hunk 4]
```diff
@@ -5,8 +5,8 @@ interface Props {
   value: string;
   isChecked: boolean;
   isSelected: boolean;
-  onMouseEnter: () => void;
   onClick: () => void;
+  onMouseMove: () => void;
```
Comment (hokolomopo) [inline src/components/filters/filter_menu_item/filter_menu_value_item.ts:9]: Wee this show that props validation would be useful after all 🙈

File: src/components/scrollbar/scrollbar_horizontal.ts [hunk 5]
```diff
@@ -54,3 +54,7 @@ export class HorizontalScrollBar extends Component<Props, SpreadsheetChildEnv> {
     });
   }
 }
+
+HorizontalScrollBar.props = {
+  position: { type: Object, optional: true },
```
Comment (hokolomopo) [inline src/components/scrollbar/scrollbar_horizontal.ts:59]: This props name is confusing. I feel liks it's a bit strange to have a position that's optional. And it's optional because that's not really the scrollbar position, but an offset to add to the viewportRect. 
Tl;dr : the props should be names offset, not position

File: src/components/scrollbar/scrollbar_vertical.ts [hunk 6]
```diff
@@ -54,3 +54,7 @@ export class VerticalScrollBar extends Component<Props, SpreadsheetChildEnv> {
     });
   }
 }
+
+VerticalScrollBar.props = {
+  position: { type: Object, optional: true },
```
Comment (hokolomopo) [inline src/components/scrollbar/scrollbar_vertical.ts:59]: Same as HorizonalScrollbar

File: src/components/side_panel/chart/gauge_chart_panel/gauge_chart_config_panel.ts [hunk 7]
```diff
@@ -47,3 +47,9 @@ export class GaugeChartConfigPanel extends Component<Props, SpreadsheetChildEnv>
     });
   }
 }
+
+GaugeChartConfigPanel.props = {
+  figureId: String,
```
Comment (hokolomopo) [inline src/components/side_panel/chart/gauge_chart_panel/gauge_chart_config_panel.ts:52]: `figureId` is only used once in all the design/config chart side panels (`LineConfigPanel`). And I'm pretty sure it's possible to replace it with props.definition in the `LineConfigPanel`.
May be worth to remove it, that's a whole 10 lines of code that we could remove 😮

**Review by hokolomopo — COMMENTED** [review]
Props validation looks a bit useful, you've already found 3-4 places with wrong props 👍

File: src/components/composer/composer/composer.ts [hunk 8]
```diff
@@ -572,3 +572,12 @@ export class Composer extends Component<Props, SpreadsheetChildEnv> {
     this.processTokenAtCursor();
   }
 }
+
+Composer.props = {
+  inputStyle: { type: String, optional: true },
+  rect: { type: Object, optional: true },
+  delimitation: { type: Object, optional: true },
+  focus: { type: String, optional: true },
```
Comment (rrahir) [inline src/components/composer/composer/composer.ts:580]: To be really thorough, we should go as far as this:
```suggestion
 focus: { validate:  v => ["inactive", "cellFocus", "contentFocus"].includes(v), optional: true },
```
  or maybe
``` typescript  
const focusValues = ["inactive", "cellFocus", "contentFocus"] as const;
  ...
interface Props {
...
focus: typeof  focusValues[number]
}

Compser.props = {
    ...,
    focus: { validate:  v => focusValues.includes(v), optional: true },
}
```
Redundant here in TS but addresses the actual validation ^^  That being said, we have to stop somewhere to avoid typing every object passed as props. it's just that this case is kind of specific and easy to implement

File: src/components/composer/grid_composer/grid_composer.ts [hunk 9]
```diff
@@ -146,3 +147,9 @@ export class GridComposer extends Component<Props, SpreadsheetChildEnv> {
     `;
   }
 }
+
+GridComposer.props = {
+  focus: String,
```
Comment (rrahir) [inline src/components/composer/grid_composer/grid_composer.ts:152]: same here

File: src/components/error_tooltip/error_tooltip.ts [hunk 10]
```diff
@@ -26,6 +26,11 @@ class ErrorToolTip extends Component<ErrorToolTipProps> {
   static components = {};
 }
 
+ErrorToolTip.props = {
+  text: String,
+  onClosed: { type: Function, optional: true },
```
Comment (rrahir) [inline src/components/error_tooltip/error_tooltip.ts:31]: add it to ErrorToolTipProps

File: src/components/filters/filter_menu_item/filter_menu_value_item.ts (hunk 4 above)
Comment (rrahir) [inline src/components/filters/filter_menu_item/filter_menu_value_item.ts:9]: useless diff or align with FilterMenuValueItem.props ?

File: src/components/menu/menu.ts [hunk 11]
```diff
@@ -258,3 +258,11 @@ export class Menu extends Component<Props, SpreadsheetChildEnv> {
     }
   }
 }
+
+Menu.props = {
+  position: Object,
+  menuItems: Array,
+  depth: { type: Number, optional: true },
```
Comment (rrahir) [inline src/components/menu/menu.ts:265]: interface says it's mandatory

File: src/components/popover/popover.ts [hunk 12]
```diff
@@ -99,3 +99,14 @@ export class Popover extends Component<Props, SpreadsheetChildEnv> {
     );
   }
 }
+
+Popover.props = {
+  position: Object,
+  marginTop: { type: Number, optional: true },
```
Comment (rrahir) [inline src/components/popover/popover.ts:105]: interface says it's mandatory, same goes for flipHorizontalOffset and flipVerticalOffset

File: src/components/side_panel/conditional_formatting/conditional_formatting.ts [hunk 13]
```diff
@@ -751,3 +751,8 @@ export class ConditionalFormattingPanel extends Component<Props, SpreadsheetChil
     this.state.rules.iconSet.icons[target] = icon;
   }
 }
+
+ConditionalFormattingPanel.props = {
+  selection: { type: Object, optional: true },
+  onCloseSidePanel: Function,
```
Comment (rrahir) [inline src/components/side_panel/conditional_formatting/conditional_formatting.ts:757]: add it to the interface

File: src/components/side_panel/custom_currency/custom_currency.ts [hunk 14]
```diff
@@ -193,3 +195,7 @@ export class CustomCurrencyPanel extends Component<any, SpreadsheetChildEnv> {
     return currency.name + (currency.code ? ` (${currency.code})` : "");
   }
 }
+
+CustomCurrencyPanel.props = {
+  onCloseSidePanel: Function,
```
Comment (rrahir) [inline src/components/side_panel/custom_currency/custom_currency.ts:200]: add it to the props. Actually, it might be nice to have some sort of props check in typescript for every component added to the sidePanelRegistry (i.e. all should have a onCloseSidepanel in its props) Don't know if it's easy to do, doesn't look like it

Comment (rrahir) [reply src/components/side_panel/custom_currency/custom_currency.ts:200]: maybe all inherit from a component w/ that props? really don't know

**Review by rrahir — COMMENTED** [review]
mostly comments about the double use of optional props and default values on several props. don't think we need both

File: src/components/figures/figure/figure.ts [hunk 15]
```diff
@@ -395,3 +395,9 @@ export class FigureComponent extends Component<Props, SpreadsheetChildEnv> {
     }
   }
 }
+
+FigureComponent.props = {
+  sidePanelIsOpen: Boolean,
```
[the remark this answers is not in the document]
Comment (pro-odoo) [reply src/components/figures/figure/figure.ts:400]: Let's keep that for now, and remove the props if needed later

File: src/components/filters/filter_menu/filter_menu.ts (hunk 3 above)
Comment (pro-odoo) [reply src/components/filters/filter_menu/filter_menu.ts:313]: Prettier accept both, I will change this one however

File: src/components/side_panel/chart/gauge_chart_panel/gauge_chart_config_panel.ts (hunk 7 above)
Comment (pro-odoo) [reply src/components/side_panel/chart/gauge_chart_panel/gauge_chart_config_panel.ts:52]: It's used a lot in Odoo, so I will check that later

File: src/components/filters/filter_menu_item/filter_menu_value_item.ts (hunk 4 above)
Comment (pro-odoo) [reply src/components/filters/filter_menu_item/filter_menu_value_item.ts:9]: The props was missing but used only in the template so typescript didn't see it

File: src/components/side_panel/custom_currency/custom_currency.ts (hunk 14 above)
Comment (pro-odoo) [reply src/components/side_panel/custom_currency/custom_currency.ts:200]: Don't know, same goes for tooltips :/
For now I let as it without any good solution

**Comment by rrahir** [comment]
@robodoo r+

**Comment by rrahir** [comment]
@pro-odoo I didn't realise it at that time but we actually need to backport this in 16.0 ...

**Comment by pro-odoo** [comment]
> @pro-odoo I didn't realise it at that time but we actually need to backport this in 16.0 ...

Why ?
